tests/resource_tests.cc

```diff
@@ -99,3 +101,45 @@ TEST_F(ResourceTest, ProductExtractGraphid) {
   EXPECT_NE(p1->state_id(), p3->state_id());
 }
 
+TEST_F(ResourceTest, DefaultPackageId) {
+  EXPECT_EQ(m1->package_id(), Package::unpackaged_id());
+  EXPECT_EQ(m2->package_id(), Package::unpackaged_id());
+  EXPECT_EQ(p1->package_id(), Package::unpackaged_id());
+  EXPECT_EQ(p2->package_id(), Package::unpackaged_id());
+
+  Product::Ptr p3 = p1->Extract(2);
+  EXPECT_EQ(p3->package_id(), Package::unpackaged_id());
+}
+
+TEST_F(ResourceTest, ChangePackageId) {
+  ctx->AddPackage("foo", 1, 5, "first");
+  Package::Ptr pkg = ctx->GetPackageByName("foo");
+  int pkg_id = pkg->id();
+
+  Product::Ptr p3 = p1->Extract(2);
+  p3->ChangePackageId(pkg_id);
+  EXPECT_EQ(p3->package_id(), pkg_id);
+  EXPECT_EQ(p1->package_id(), Package::unpackaged_id());
+
+  m1->ChangePackageId(pkg_id);
+  EXPECT_EQ(m1->package_id(), pkg_id);
+}
+
+TEST_F(ResourceTest, PackageResource) {
+  ctx->AddPackage("foo", 1, 5, "first");
+  Package::Ptr pkg = ctx->GetPackageByName("foo");
+  int pkg_id = pkg->id();
+
+  // nothing packaged
+  Product::Ptr p3 = p1->Extract(0.5);
+  std::vector<Product::Ptr> p3_pkgd = p3->Package<Product>(pkg);
+
+  // everything stays in old product, with same (default) package id
+  EXPECT_EQ(p3->package_id(), Package::unpackaged_id());
+  EXPECT_EQ(p3->quantity(), 0.5);
+
+  // all packaged
+  std::vector<Product::Ptr> p1_pkgd = p1->Package<Product>(pkg);
+  EXPECT_EQ(p1->quantity(), 0);
+  EXPECT_EQ(p1_pkgd[0]->package_id(), pkg_id);
```

Comment:
should we also check the number of packages?  If I read this right, `p1_pkgd` should be length one: all of its 3 units of quantity in a single package.

We could also package `p2` and its 7 units would end up in two packages of size 5 and 2

Reply:
you'll never guess what was in my stashes.... though I didn't I have the two packages case